libusb/os/windows_hotplug.c

```diff
@@ -0,0 +1,307 @@
+ * 2. Run get_device_list which will set devices still connected to UNCHANGED and new devices to ARRIVED.
+ * Since the marker is not modified for devices which had disconnected and are thus no longer enumerated,
+ * all devices have appropriate connection status and the end of the process */
+static int windows_get_device_list(struct libusb_context *ctx)
+{
+	// note: context device list is protected by active_contexts_lock
+	struct libusb_device *dev;
+	for_each_device(ctx, dev)
+	{
+		((struct winusb_device_priv *)usbi_get_device_priv(dev))->hotplug_status = LEFT;
+	}
+
+	return ((struct windows_context_priv *)usbi_get_context_priv(ctx))->backend->get_device_list(ctx);
+}
+
+int windows_initial_scan_devices(struct libusb_context *ctx)
+{
+	usbi_mutex_static_lock(&active_contexts_lock);
+	
+	int ret =  windows_get_device_list(ctx);
+	
+	usbi_mutex_static_unlock(&active_contexts_lock);
+	return ret;
+}
+
+static void windows_refresh_device_list(struct libusb_context *ctx)
+{
+	int ret = windows_get_device_list(ctx);
+	if(ret != LIBUSB_SUCCESS)
+	{
+		usbi_err(ctx, "hotplug failed to retrieve current list with error: %s", libusb_error_name(ret)); 
+		return;
+	}
+
+	struct libusb_device *dev, *next_dev;
+	struct winusb_device_priv *priv;
+
+	for_each_device_safe(ctx, dev, next_dev)
+	{
+		priv = usbi_get_device_priv(dev);
+		if(priv->hotplug_status != LEFT)
+		{
+			continue;
+		}
+
+		if(priv->initialized)
+		{
+			usbi_disconnect_device(dev);			
+		}
+		else
+		{
+			usbi_detach_device(dev);	
+		}
+	}
+
+	for_each_device(ctx, dev)
+	{
+		priv = usbi_get_device_priv(dev);
+
+		if(priv->hotplug_status != ARRIVED)
+		{
+			continue;
+		}
+
+		usbi_hotplug_notification(ctx, dev, LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED);
+	}
+}
+
+static void windows_refresh_device_list_for_all_ctx(void)
+{
+	usbi_mutex_static_lock(&active_contexts_lock);
+
+	struct libusb_context *ctx;
+	for_each_context(ctx)
+	{
+		windows_refresh_device_list(ctx);
+	}
+	usbi_mutex_static_unlock(&active_contexts_lock);
+}
+
+#define WND_CLASS_NAME TEXT("libusb-1.0-windows-hotplug")
+
+static bool init_wnd_class(void)
+{
+	WNDCLASS wndClass = { 0 };
+	wndClass.lpfnWndProc = windows_proc_callback;
+	wndClass.hInstance = GetModuleHandle(NULL);
+	wndClass.lpszClassName = WND_CLASS_NAME;
+
+	if (!RegisterClass(&wndClass)) 
+	{
+		log_error("event thread: RegisterClass");
+		return false;
+	}
+
+	return true;
+}
+
+static DWORD WINAPI windows_event_thread_main(LPVOID lpParam)
+{
+	UNUSED(lpParam);
+
+	usbi_dbg(NULL, "windows event thread entering");
+
+	if (!init_wnd_class())
+	{
+		return (DWORD)-1;
+	}
+
+	windows_event_hWnd = CreateWindow(
+		WND_CLASS_NAME,
+		TEXT(""),
+		0, 
+		0, 0, 0, 0,
+		NULL, NULL,
+		GetModuleHandle(NULL),
+		NULL);
+
+	if (windows_event_hWnd == NULL)
+	{
+		log_error("event thread: CreateWindow");
+		return (DWORD)-1;
+	}
+
+	MSG msg;
+	BOOL ret_val;
+
+	while ((ret_val = GetMessage(&msg, windows_event_hWnd, 0, 0)) != 0)
+	{
+		if (ret_val == -1)
```

Comment:
Is there no macro for this value?

Reply:
Well, short answer is "no", unfortunately. This GetMessage function is ... "special" ;-) It returns a BOOL, but sometimes also -1. See here: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getmessage , where one can read this:
"
Return value
Type: BOOL

If the function retrieves a message other than [WM_QUIT](https://learn.microsoft.com/en-us/windows/desktop/winmsg/wm-quit), the return value is nonzero.

If the function retrieves the [WM_QUIT](https://learn.microsoft.com/en-us/windows/desktop/winmsg/wm-quit) message, the return value is zero.

**If there is an error, the return value is -1**. For example, the function fails if hWnd is an invalid window handle or lpMsg is an invalid pointer. To get extended error information, call [GetLastError](https://learn.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror).

Because the return value can be nonzero, zero, or -1, avoid code like this:


`while (GetMessage( lpMsg, hWnd, 0, 0)) ...`
The possibility of a -1 return value in the case that hWnd is an invalid parameter (such as referring to a window that has already been destroyed) means that such code can lead to fatal application errors. Instead, use code like this:


```
BOOL bRet;

while( (bRet = GetMessage( &msg, hWnd, 0, 0 )) != 0)
{ 
    if (bRet == -1)
    {
        // handle the error and possibly exit
    }
    else
    {
        TranslateMessage(&msg); 
        DispatchMessage(&msg); 
    }
}
```
I tried to conform to best practices, even if it looks strange...

Reply:
Thanks for the explanation. It was "special", yes.